src/Conversion/ONNXToTOSA/Tensor/Resize.cpp

```diff
@@ -0,0 +1,314 @@
+  ///
+  /// ### align_corners ###
+  /// - scale_x_d = (input_size - 1)
+  /// - scale_x_n = (output_size - 1)
+  /// - offset = 0
+  ///
+  /// ### asymmetric ###
+  /// - scale_x_d = input_size
+  /// - scale_x_n = output_size
+  /// - offset = 0
+  ///
+  /// ## nearest_mode ##
+  /// If mode == nearest, then ONNX can set the nearest_mode attribute.
+  /// The standard case for TOSA of round_half_up. Support for floor can
+  /// be achieved with:
+  /// offset_x -= n/2
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    auto resizeOp = llvm::cast<ONNXResizeOp>(op);
+    Location loc = op->getLoc();
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+
+    TosaBuilder tosaBuilder(rewriter, loc);
+
+    Value input = adaptor.getX();
+    auto inputType = input.getType().dyn_cast<RankedTensorType>();
+
+    auto resultType =
+        resizeOp.getResult().getType().dyn_cast<RankedTensorType>();
+
+    StringRef mode = adaptor.getMode();
+    StringRef nearestMode = adaptor.getNearestMode();
+    StringRef coordinateTransformationMode =
+        adaptor.getCoordinateTransformationMode();
+    std::optional<ArrayAttr> axis = adaptor.getAxes();
+    int64_t antialias = adaptor.getAntialias();
+
+    if (!inputType || inputType.getRank() != 4) {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA only support 4D tensors as input of resize.");
+    }
+
+    // With only static dimensions, scales and sizes as inputs are not relevant
+    // anymore.
+    if (inputType.isDynamicDim(2) || inputType.isDynamicDim(3)) {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "Only static sized tensors are supported.");
+    }
+
+    if (mode == "cubic") {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA does not support cubic interpolation.");
+    }
+
+    if (mode == "nearest" &&
+        (nearestMode == "ceil" || nearestMode == "round_prefer_floor")) {
+      return rewriter.notifyMatchFailure(resizeOp,
+          "TOSA does not support ceil and round_prefer_floor as nearestMode.");
+    }
+
+    // This also makes roi as an input irrelevant.
+    if (coordinateTransformationMode == "tf_crop_and_resize") {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA does not support tf_crop_and_resize.");
+    }
+
+    if (antialias != 0)
+      return rewriter.notifyMatchFailure(
+          op, "TOSA does not support antialiasing.");
+
+    auto inputShape = inputType.getShape();
+    auto outputShape = resultType.getShape();
+
+    if (inputShape[0] != outputShape[0] || inputShape[1] != outputShape[1])
+      return rewriter.notifyMatchFailure(
+          op, "Cannot resize non spatial dimensions.");
+
+    // Get axis values if set. Default is all axis in normal order.
+    llvm::SmallVector<int64_t> axisVec;
+    if (axis.has_value()) {
+      valuesFromAxis(&axis.value(), axisVec);
```

Comment:
I wonder how strict we should be here because behavior is undefined if an axis is repeated and the ONNX Resize verifier does not check for that particular case. What do you think?

Reply:
Hmm I would say this kind of checking should happen in the verifier and not in the conversion. The verifiers in onnx-mlir are not super mature, maybe we should invest in them if we see a failure that could have been noticed earlier